### PR TITLE
Associate binding secret to a custom schema path in the workload

### PR DESCRIPTION
**Pull request by sbose78**
### Motivation

- Many Application workloads do not support podSpec, and might need association of the binding secret in an arbitrary location.
- Since we are really concerned about `containers` and `volumes`, users should be able to provide a custom path to them as well.

Example, in the below workload, the binding secret has to be associated in 
`spec.eventSourceSecret` instead of `spec.template.spec.containers`

```
apiVersion: sources.eventing.knative.dev/v1alpha1
 kind: KafkaSource
 metadata:
  name: my-kafka-source
 spec:
  consumerGroup: knative-group
  eventSourceSecret: eventsource-binding-foo-bar-binding-secret <--- Association made by service binding operator
  sink:
    ref:
      apiVersion: serving.knative.dev/v1alpha1
      kind: Service
      name: event-display
```

To accomplish the same, the `ServiceBindingRequest`'s  `applicationSelector` could look like the following:

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest

metadata: 
  name: eventsource-binding-foo-bar
  namespace: service-binding-demo

spec: 
  applications: 
    - bindingPath:
         customSecretPath: spec.eventSourceSecret # <--- In  KafkaSource
      group: sources.eventing.knative.dev
      resource: kafkasources
      resourceRef: my-kafka-source
      version: v1alpha1

  services:                                                                                                         
    - group: kafka.strimzi.io
      version: v1beta1
      kind: KafkaTopic
      resourceRef: my-kafka-topic 
    - group: kafka.strimzi.io
      version: v1beta1
      kind: Kafka
      resourceRef: my-kafka-cluster
```

### Changes

New non-mandatory API field `bindingPath` introduced.

```
applications:
  - bindingPath:
       podSpecPath:
         containers: spec.template.spec.containers
         volumes: spec.template.spec.volumes
       customSecretPath: spec.custom.path
     group: apps.openshift.io
     resource: deploymentconfigs
     resourceRef: nodejs-rest-http-crud
     version: v1
```

**Default behaviour**
If `bindingPath` is not set, default containers/volumes path is used as per podSpec.

### Testing
Ensured that binding is successful with/without podSpec path set.

### CODE REVIEW

**Comment by sbose78** [comment]
CC @arthurdm @pdettori , thoughts?

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go [hunk 1]
```diff
@@ -71,6 +71,35 @@ type ApplicationSelector struct {
...
+	// where the binding workload would be referenced.
+	// +optional
+	BindingPath *BindingPath `json:"bindingPath,omitempty"`
+}
+
+// BindingPath defines the path to the field where the binding would be
+// embedded in the workload
+type BindingPath struct {
+	// PodSpecPath overrides the default podSpec path
+	// optional
```
Comment (Avni-Sharma) [inline pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: nit pick `+` with optional

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go [hunk 2]
```diff
@@ -71,6 +71,35 @@ type ApplicationSelector struct {
...
+
+// PodSpecPath overrides the default podSpec path
+type PodSpecPath struct {
+	// Containers defines the path to the corev1.Containers reference
+	// Example: "spec.template.spec.containers"
+	Containers string `json:"containers"`
+
+	// Containers defines the path to the corev1.Volumes reference
+	// Example: "spec.template.spec.volumes"
+	Volumes string `json:"volumes"`
```
Comment (Avni-Sharma) [inline pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: Shouldn't `volumes` be optional?

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go (hunk 1 above)
Comment (sbose78) [reply pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: Right 2a47c75 , thank you.

File: pkg/apis/apps/v1alpha1/servicebindingrequest_types.go (hunk 2 above)
Comment (sbose78) [reply pkg/apis/apps/v1alpha1/servicebindingrequest_types.go]: Good question: 

( thinking out aloud )
If a user is specifying a custom path to `containers`, is it likely that the user might also want to specify a custom path to `volumes` ?

Would a user want to skip `volumes` altogether? In that case, what's the best way to express that? I'm not too sure. 

This part of the API might need some evolution.

**Comment by arthurdm** [comment]
sounds like a good addition @sbose78  👍 

One of the advantages of the `applicationSelector`'s toggle to turn off is this - since the workload may have special considerations that its deployment Operator knows about (e.g. where to bind the secret, how to bind it, etc).   I do agree though, for workoads that want the SBO to do the binding, this enhancement is needed.

**Comment by sbose78** [comment]
Right, this is for the scenario where the application deployer is OK to have her object patched.

**Comment by pdettori** [comment]
> CC @arthurdm @pdettori , thoughts?

@sbose78 I think `customSecretPath` is a valuable feature. I think though that there might be situations where not everything needed to connect is retrieved from a single secret. For example, the KafkaSource in your example above requires also [BootstrapServers](https://github.com/knative/eventing-contrib/blob/master/kafka/source/pkg/apis/sources/v1alpha1/kafka_types.go#L104) which might not be a fixed value (for example, we might be using an in-cluster or off-cluster Kafka - the latter with the IBM Cloud Operator.

**Comment by sbose78** [comment]
cc @matzew  @lberk  @ianpartridge , Thoughts?

**Comment by otaviof** [comment]
@sbose78, to follow up on #366 [comment](https://github.com/redhat-developer/service-binding-operator/pull/366#pullrequestreview-370830366), should the `ServiceBindingRequest` on this PR look like the following:

```yml
spec:
  [...]
  binding:
    paths:
      containers: spec.template.spec.containers
      volumes: spec.template.spec.volumes 
    group: ''
    kind: ConfigMap
    version: v1
```

Notice `spec.binding.paths`, it should come with default values based on `group`, `version` and `kind` specified for the binding, where the user could then overwrite.

**Comment by sbose78** [comment]
Looks good @otaviof , added the custom secret path as `binding.paths.secret` below:

```
binding:
    paths:
      secret: spec
      containers: spec.template.spec.containers
      volumes: spec.template.spec.volumes 
```

**Comment by sbose78** [comment]
Instead of `binding,paths.secret`, I would go for 

```
binding:
    paths:
      data: spec.credentials
      containers: spec.template.spec.containers
      volumes: spec.template.spec.volumes 
```

to be able to support https://github.com/redhat-developer/service-binding-operator/pull/366

File: pkg/controller/servicebindingrequest/binding.go [hunk 3]
```diff
@@ -339,6 +339,15 @@ func BuildServiceBinder(options *ServiceBinderOptions) (*ServiceBinder, error) {
 	// gather related secret, again only appending it if there's a value.
 	secret := NewSecret(options.DynClient, plan)
 
+	if options.SBR.Spec.ApplicationSelector.BindingPath == nil {
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/binding.go]: From what I understood always apply the `PodSpecPath` binding if `BindingPath` is not declared, is it right?

File: pkg/controller/servicebindingrequest/binder.go [hunk 4]
```diff
@@ -24,10 +26,15 @@ import (
...
-	volumesPath = []string{"spec", "template", "spec", "volumes"}
+	// containersPath has the logical path logical path
+	// to find containers on supported objects
+	// Used as []string{"spec", "template", "spec", "containers"}
+
+	pathToContainers = "spec.template.spec.containers"
+
+	// logical path to find volumes on supported objects
+	// used as []string{"spec", "template", "spec", "volumes"}
+	pathToVolumes = "spec.template.spec.volumes"
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/binder.go]: Both `pathToContainers` and `pathToVolumes` are default values for two specific properties in `PodSpecPath`.

Perhaps `DefaultPodSpecContainersPath` and `DefaultPodSpecVolumesPath` would make it more clear?

File: pkg/controller/servicebindingrequest/binding_test.go [hunk 5]
```diff
@@ -148,6 +147,35 @@ func TestServiceBinder_Bind(t *testing.T) {
...
 				}
 			}
+
+			providedBindingPath := args.options.SBR.Spec.ApplicationSelector.BindingPath
+			computedBindingPath := sb.SBR.Spec.ApplicationSelector.BindingPath
+
+			if providedBindingPath != nil {
+
+				// user has provided BindingPath information explicitly
+				require.True(t, providedBindingPath.PodSpecPath != nil ||
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/binding_test.go]: I assume this is related to the injection of `BindingPath` in `BuildServiceBinder`, and perhaps should be moved to a separate function specialized in validating `BindingPath` values.

In this case, `BuildServiceBinder` could return an error (let's say `InvalidBindingPathErr`) and specified in `wantBuildErr`.

The validation function could be properly tested to cover `BindingPath` semantics and the contract would be the error returned by `BuildServiceBinder`.

File: pkg/controller/servicebindingrequest/binder.go (hunk 4 above)
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder.go]: Agreed.

File: pkg/controller/servicebindingrequest/binding.go (hunk 3 above)
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binding.go]: Yeah. Correct.

Also, if the user specifies the path to the secret explicitly, I am expecting the user to also specify the containersPath and the volumesPaths , ie, the podSpecPath, explicitly. 

What do you think?

File: pkg/controller/servicebindingrequest/binding_test.go (hunk 5 above)
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binding_test.go]: Sure, will do!

**Comment by ZhuangYuZY** [comment]
@sbose78 Hello, when the PR can get merged, we will need to it to bind secret with custom resource. Thank you.

**Comment by cdlliuy** [comment]
@sbose78 ,  we rebased and further edit this PR to resolve current conflicting internally, and validated it works in our scenario.    
Would you like to do the rebase and get it merged recently ?   We hope to get this feature as early as possible. Thanks a lot !

**Comment by sbose78** [comment]
Thanks, I'll clean this up. Do you mind sharing details on your scenario ?

**Comment by ZhuangYuZY** [comment]
@sbose78 here is the scenario we want to have https://github.com/redhat-developer/service-binding-operator/issues/400#issuecomment-615258348
Thank you for your help.

**Comment by sbose78** [comment]
@cdlliuy if you have things sorted out locally, do you mind making a PR to my branch ( unless I get to it before you do :) )?

**Comment by cdlliuy** [comment]
@sbose78  sure.  @qibobo  will follow up on this.

**Comment by sbose78** [comment]
Conficts resolved.

File: pkg/controller/servicebindingrequest/binder_test.go [hunk 6]
```diff
@@ -102,6 +113,24 @@ func TestBinderNew(t *testing.T) {
 		require.Equal(t, "value", list[0].Value)
 	})
 
+	t.Run("update custom secret path ", func(t *testing.T) {
+		list, err := binder.search()
```
[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder_test.go:121]: That is indeed the one being used 
```
customSecretPath := "metadata.clusterName"
	sbr.Spec.ApplicationSelector.BindingPath = &v1alpha1.BindingPath{
		PodSpecPath: &v1alpha1.PodSpecPath{
			Containers: pathToContainers,
			Volumes:    pathToVolumes,
		},
		CustomSecretPath: &customSecretPath,
	}
	f.AddMockedUnstructuredDeployment("ref", matchLabels)

	binder := NewBinder(
		context.TODO(),
		f.FakeClient(),
		f.FakeDynClient(),
		sbr,
		[]string{},
	)
```

[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder_test.go:121]: Ah OK, I see your previous comment.

File: pkg/controller/servicebindingrequest/binder_test.go [hunk 7]
```diff
@@ -39,6 +41,15 @@ func TestBinderNew(t *testing.T) {
...
 	sbr := f.AddMockedServiceBindingRequest(name, "ref", "", deploymentsGVR, matchLabels)
+
+	customSecretPath := "metadata.clusterName"
+	sbr.Spec.ApplicationSelector.BindingPath = &v1alpha1.BindingPath{
+		PodSpecPath: &v1alpha1.PodSpecPath{
+			Containers: pathToContainers,
+			Volumes:    pathToVolumes,
+		},
+		CustomSecretPath: &customSecretPath,
+	}
```
[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder_test.go:104]: done 468641292ef4a7b1851ce33cbf1347f8606d812e

File: pkg/controller/servicebindingrequest/binder_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder_test.go]: Done 468641292ef4a7b1851ce33cbf1347f8606d812e

File: pkg/controller/servicebindingrequest/binder.go (hunk 4 above)
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binder.go]: Done 468641292ef4a7b1851ce33cbf1347f8606d812e

File: pkg/controller/servicebindingrequest/binding_test.go (hunk 5 above)
Comment (sbose78) [reply pkg/controller/servicebindingrequest/binding_test.go]: Did my best here b142835763bd2e3f68b64191a3b8427d7d41ee77 . Let me know!

**Comment by sbose78** [comment]
@cdlliuy @qu1queee  May I request a review from you folks, please ?

**Comment by qibobo** [comment]
There seems to be a problem that after SBR is deleted, the secret injection can not be delete from the application. Below is what I have done for the test:

**1 create a CRD**
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: appconfigs.stable.example.com
spec:
  group: stable.example.com
  versions:
    - name: v1
      served: true
      storage: true
  scope: Namespaced
  names:
    plural: appconfigs
    singular: appconfig
    kind: AppConfig
    shortNames:
    - ac
```

**2 create a CR:**

```
apiVersion: "stable.example.com/v1"
kind: AppConfig
metadata:
  name: demo-appconfig
spec:
  uri: "some uri"
  Command: "some command"
  image: my-image
  spec:
    containers:
    - name: hello-world
      # Image from dockerhub, This is the import path for the Go binary to build and run.
      image: yusufkaratoprak/kubernetes-gosample:latest
      ports:
      - containerPort: 8090
```
And we can see there is no secret injected:

```
kubectl get appconfigs.stable.example.com demo-appconfig -o yaml
apiVersion: stable.example.com/v1
kind: AppConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"stable.example.com/v1","kind":"AppConfig","metadata":{"annotations":{},"name":"demo-appconfig","namespace":"default"},"spec":{"Command":"some command","image":"my-image","spec":{"containers":[{"image":"yusufkaratoprak/kubernetes-gosample:latest","name":"hello-world","ports":[{"containerPort":8090}]}]},"uri":"some uri"}}
  creationTimestamp: "2020-04-22T09:15:51Z"
  generation: 1
  name: demo-appconfig
  namespace: default
  resourceVersion: "1496884"
  selfLink: /apis/stable.example.com/v1/namespaces/default/appconfigs/demo-appconfig
  uid: dc3aee00-9258-4017-8a61-a1ca9dd1f6e7
spec:
  Command: some command
  image: my-image
  spec:
    containers:
    - image: yusufkaratoprak/kubernetes-gosample:latest
      name: hello-world
      ports:
      - containerPort: 8090
  uri: some uri
```

**3 create a SBR:**

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-sample
spec:
  envVarPrefix: qiye111
  applicationSelector:
    resourceRef: demo-appconfig
    group: stable.example.com
    version: v1
    resource: appconfigs 
    bindingPath:
      podSpecPath:
        containers: spec.spec.containers
        volumes: spec.spec.volumes
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator3
      id: zzz
      envVarPrefix: qiye
  customEnvVar:
    - name: VCAP_SERVICES
      value: '{"language-translator":[{"credentials":{{json .zzz.status.secretName}},"name":"thetranslator3","plan":"standard"}]}'
    - name: TRANSLATOR_NAME
      value: "thetranslator3"
```

So we can find the secret ` binding-request-sample` has been injected:

```
kubectl get appconfigs.stable.example.com demo-appconfig -o yaml
apiVersion: stable.example.com/v1
kind: AppConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"stable.example.com/v1","kind":"AppConfig","metadata":{"annotations":{},"name":"demo-appconfig","namespace":"default"},"spec":{"Command":"some command","image":"my-image","spec":{"containers":[{"image":"yusufkaratoprak/kubernetes-gosample:latest","name":"hello-world","ports":[{"containerPort":8090}]}]},"uri":"some uri"}}
  creationTimestamp: "2020-04-22T09:15:51Z"
  generation: 2
  name: demo-appconfig
  namespace: default
  resourceVersion: "1496926"
  selfLink: /apis/stable.example.com/v1/namespaces/default/appconfigs/demo-appconfig
  uid: dc3aee00-9258-4017-8a61-a1ca9dd1f6e7
spec:
  Command: some command
  image: my-image
  spec:
    containers:
    - env:
      - name: ServiceBindingOperatorChangeTriggerEnvVar
        value: "2020-04-22T09:16:20Z"
      envFrom:
      - secretRef:
          name: binding-request-sample
      image: yusufkaratoprak/kubernetes-gosample:latest
      name: hello-world
      ports:
      - containerPort: 8090
      resources: {}
  uri: some uri
```

**4 delete the SBR**
`servicebindingrequest.apps.openshift.io "binding-request-sample" deleted`

But the secret is still here:

```
kubectl get appconfigs.stable.example.com demo-appconfig -o yaml
apiVersion: stable.example.com/v1
kind: AppConfig
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"stable.example.com/v1","kind":"AppConfig","metadata":{"annotations":{},"name":"demo-appconfig","namespace":"default"},"spec":{"Command":"some command","image":"my-image","spec":{"containers":[{"image":"yusufkaratoprak/kubernetes-gosample:latest","name":"hello-world","ports":[{"containerPort":8090}]}]},"uri":"some uri"}}
  creationTimestamp: "2020-04-22T09:15:51Z"
  generation: 2
  name: demo-appconfig
  namespace: default
  resourceVersion: "1496926"
  selfLink: /apis/stable.example.com/v1/namespaces/default/appconfigs/demo-appconfig
  uid: dc3aee00-9258-4017-8a61-a1ca9dd1f6e7
spec:
  Command: some command
  image: my-image
  spec:
    containers:
    - env:
      - name: ServiceBindingOperatorChangeTriggerEnvVar
        value: "2020-04-22T09:16:20Z"
      envFrom:
      - secretRef:
          name: binding-request-sample
      image: yusufkaratoprak/kubernetes-gosample:latest
      name: hello-world
      ports:
      - containerPort: 8090
      resources: {}
  uri: some uri
```

**Comment by sbose78** [comment]
> servicebindingrequest.apps.openshift.io "binding-request-sample" deleted

Good feedback, the deletion probably still assumes that the default podspec paths.

**Comment by sbose78** [comment]
Would be nice if someone could lend me a helping hand to resolve the remaining issue here ( PRs to my branch welcome ). I'm off for the next couple of days.

**Comment by qibobo** [comment]
@sbose78 
I can help it.
Besides the deletion problem, is there any other issues need to fixed?

**Comment by DhritiShikhar** [comment]
@qibobo Shoubhik is on leave as he is sick. 
I am not sure if he'll be able to respond.
I can see that other review comments are already resolved.
So, Let's go ahead and fix the deletion problem and get this PR merged?

**Comment by qibobo** [comment]
@DhritiShikhar 
Ok. I will submit the fix to his branch.

**Comment by qibobo** [comment]
@sbose78 @DhritiShikhar 
I finally find that the issue I reported above is not a code issue. In my test I used an old version of SBR CRD, so the bind process does not really complete, just inject the secret, but no update SBR status and add finalizer.
So I am OK with the PR now.

**Comment by DhritiShikhar** [comment]
@qibobo Okay Thanks for the update.
We will merge this PR.

**Comment by isutton** [comment]
I'm not sure whether having `bindingPath` in `applicationSelector` is the right approach when multiple applications are involved; I shared some notes I've collected which lead to a slightly different design in a Gist: https://gist.github.com/isutton/12a20a8aa1f9a1469a963caa4fc13f3a.

I want to avoid having to re-work this feature if/when multiple applications are merged into master.

**Comment by isutton** [comment]
I'm holding this merge as I'd like to see more discussion going on regarding whether we'll support multiple applications and the implication of this PR with it.

/hold

**Comment by sbose78** [comment]
I'm good with this 

```
- type: PodSpecSlice
  path: .spec.template.spec.containers
  resources:
    - deployments.apps/v1
    - replicasets.core/v1
```

from here.

**Comment by arthurdm** [comment]
In the service binding spec we assume the binding request applies to just a single application.  This support the microservice model, where each microservice has an independent / self-contained set of artifacts.  

Having two or more applications share the same ServiceBindingRequest would go against that thought - i.e. you now create a monolith binding where things for a particular microservice can't grow without affecting the others.

**Comment by sbose78** [comment]
In all fairness, I now do see an impact of asking for multiple apps in a single SBR.

We wouldn't lose any functionality as such if we supported only a single app.

However we already support label selectors, do you suggest we deprecate that?

Arthur, Igor, would love to hear your opinions on that.

**Comment by sbose78** [comment]
> However we already support label selectors, do you suggest we deprecate that?

Ignore.
Label selectors are always on the same GVK.

**Comment by sbose78** [comment]
> This approach is likely to not work, since it can’t be assumed that an application matching matchLabels will 1) have declared the field in the object and 2) it is not guaranteed that matchLabels would return a list of only Deployment like objects.

Revisited your statement @isutton 

>  it is not guaranteed that matchLabels would return a list of only Deployment like objects.

It is guaranteed since matchLabels is specified per application and hence would return homogenous objects only. 

>  have declared the field in the object 

Didn't quite get this part.

**Comment by isutton** [comment]
> > This approach is likely to not work, since it can’t be assumed that an application matching matchLabels will 1) have declared the field in the object and 2) it is not guaranteed that matchLabels would return a list of only Deployment like objects.
> 
> Revisited your statement @isutton
> 
> > it is not guaranteed that matchLabels would return a list of only Deployment like objects.
> 
> It is guaranteed since matchLabels is specified per application and hence would return homogenous objects only.

The limitation happens because the GVR is being used in the client, not because the user has the intention of only using a specific application kind (`Deployment`, `DeploymentConfig` or another similar resource). 

My interpretation has been that either labels are matched and a heterogeneous set of matching objects to be inspected and modified, or a reference to a specific resource, but that seems to not be the case.

Perhaps we should revisit the history behind `matchLabels` and how it is expected to be used.

> > have declared the field in the object
> 
> Didn't quite get this part.

The field specified by `spec.applicationSelector.bindingSecretPath` must exist in the resolved objects.

**Comment by sbose78** [comment]
The labels are to be matched **only** with the specific GVK and not all kinds. Hence, the objects returned for that application will always have homogenous objects.

**Comment by cdlliuy** [comment]
@sbose78 ,  a little bit curious that why the changes in 353 will be affected by matchlabel or multiple application ?  The proposed change is to add a bindingpath under one of the applicationSelector,   which should be fine as well when expand to multiple application as well.
```
  applicationSelector:
    resourceRef: demo-appconfig
    group: stable.example.com
    version: v1
    resource: appconfigs 
    bindingPath:
      podSpecPath:
        containers: spec.spec.containers
        volumes: spec.spec.volumes
```
In the meanwhile, why multiple application?

**Comment by sbose78** [comment]
You are right, @cdlliuy. It isn't impacted. I'm waiting on approvals on this PR :) 
@Avni-Sharma @baijum @isutton @DhritiShikhar ?

**Comment by sbose78** [comment]
Folks @redhat-developer/service-binding , @ZhuangYuZY @cdlliuy is looking for an update on this :)

**Comment by sbose78** [comment]
Could you help me with the Group Version Kind of the `application` @cdlliuy ?

**Comment by cdlliuy** [comment]
@sbose78 Glad to help, but can you explain a little bit what kinds of help is needed?

**Comment by sbose78** [comment]
What is the GVK you are using @cdlliuy ?

**Comment by qibobo** [comment]
@sbose78 
Below are 2 SBR samples we are using:

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-vcap
spec:
  envVarPrefix: qiye111
  applicationSelector:
    resourceRef: mytranslator-web-knative-vcap
    group: serving.knative.dev 
    version: v1 
    resource: services 
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator
      id: zzz
      envPrefix: qiye
  customEnvVar:
    - name: VCAP_SERVICES
      value: '{{json .}}'
    - name: TRANSLATOR_NAME
      value: "thetranslator"
```


```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: batchsamplesbr
spec:
  applicationSelector:
    resourceRef: batchsample
    group: coligo.cloud.ibm.com
    version: v1alpha1
    resource: jobdefinitions
    bindingPath:
      podSpecPath:
        containers: spec.containers
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thecos2
      id: cos2
```

**Comment by Avni-Sharma** [comment]
@qibobo what is the value in 
```
  customEnvVar:
    - name: VCAP_SERVICES
      value: '{{json .}}'
```

**Comment by qibobo** [comment]
> @qibobo what is the value in
> 
> ```
>   customEnvVar:
>     - name: VCAP_SERVICES
>       value: '{{json .}}'
> ```

It means to json format all the content of secret generated by SBR.

**Comment by Avni-Sharma** [comment]
@qibobo can you please run the example again on this branch and paste the output here.

**Comment by qibobo** [comment]
@sbose78 
There are conflicts, could you fix them?

**Comment by Avni-Sharma** [comment]
> @sbose78
> There are conflicts, could you fix them?

Hi @qibobo, conflicts have been resolved. Let us know if the examples/particular use cases work on this branch.

**Comment by qibobo** [comment]
@sbose78 @Avni-Sharma 

Just verified this PR on my cluster. The feature is OK.
Below are my test steps:

**1 Test it can work with a Deployment:**
Create a deployment:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: mytranslator-web-vcap5
spec:
  replicas: 1
  selector:
    matchLabels:
      app: hello
  template:
    metadata:
      name: hello
      labels:
        app: hello
    spec:
      containers:
      - name: hello-world
        image: docker.io/qibobo/helloworld4go:latest
        ports:
        - containerPort: 8080
```
Create the SBR:

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: sbr5
spec:
  envVarPrefix: globalprefix
  applicationSelector:
    resourceRef: mytranslator-web-vcap5
    group: apps
    version: v1
    resource: deployments 
    bindingPath:
      podSpecPath:
        containers: spec.template.spec.containers
        volumes: spec.template.spec.volumes
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator5
      envVarPrefix: prefix1
  customEnvVar:
    - name: TRANSLATOR_NAME
      value: "thetranslator5"
    - name: ALL_SECRET
      value: '{{json .}}'
```
Check the secret `sbr5` has been injected:

```
kubectl get deployment mytranslator-web-vcap5 -o yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "2"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"name":"mytranslator-web-vcap5","namespace":"default"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"hello"}},"template":{"metadata":{"labels":{"app":"hello"},"name":"hello"},"spec":{"containers":[{"image":"docker.io/qibobo/helloworld4go:latest","name":"hello-world","ports":[{"containerPort":8080}]}]}}}}
  creationTimestamp: "2020-05-22T08:14:49Z"
  generation: 2
  name: mytranslator-web-vcap5
  namespace: default
  resourceVersion: "4909969"
  selfLink: /apis/apps/v1/namespaces/default/deployments/mytranslator-web-vcap5
  uid: a3486510-c81e-45ea-a1de-cd5d1bf72e21
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: hello
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: hello
      name: hello
    spec:
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-05-22T08:15:29Z"
        envFrom:
        - secretRef:
            name: sbr5
        image: docker.io/qibobo/helloworld4go:latest
        imagePullPolicy: Always
        name: hello-world
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-05-22T08:14:52Z"
    lastUpdateTime: "2020-05-22T08:14:52Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2020-05-22T08:14:50Z"
    lastUpdateTime: "2020-05-22T08:15:32Z"
    message: ReplicaSet "mytranslator-web-vcap5-674c575795" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 2
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```
Delete the SBR and secret has been removed.
```
spec:
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-05-22T08:15:29Z"
        image: docker.io/qibobo/helloworld4go:latest
        imagePullPolicy: Always
        name: hello-world
        ports:
        - containerPort: 8080
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
```

**2 Test it with a CRD:**
Create the CRD:
```
apiVersion: apiextensions.k8s.io/v1beta1
kind: CustomResourceDefinition
metadata:
  name: appconfigs.stable.example.com
spec:
  group: stable.example.com
  versions:
    - name: v1
      served: true
      storage: true
  scope: Namespaced
  names:
    plural: appconfigs
    singular: appconfig
    kind: AppConfig
    shortNames:
    - ac
```

Create the application of the CRD:

```
apiVersion: "stable.example.com/v1"
kind: AppConfig
metadata:
  name: demo-appconfig
spec:
  uri: "some uri"
  Command: "some command"
  image: my-image
  spec:
    containers:
    - name: hello-world
      # Image from dockerhub, This is the import path for the Go binary to build and run.
      image: yusufkaratoprak/kubernetes-gosample:latest
      ports:
      - containerPort: 8090
```

Create the SBR with a custom bind path:

```
bindingPath:
      podSpecPath:
        containers: spec.spec.containers
        volumes: spec.spec.volumes
```

```
apiVersion: apps.openshift.io/v1alpha1
kind: ServiceBindingRequest
metadata:
  name: binding-request-sample
spec:
  envVarPrefix: qiye111
  applicationSelector:
    resourceRef: demo-appconfig
    group: stable.example.com
    version: v1
    resource: appconfigs 
    bindingPath:
      podSpecPath:
        containers: spec.spec.containers
        volumes: spec.spec.volumes
  backingServiceSelectors:
    - group: ibmcloud.ibm.com
      version: v1alpha1
      kind: Binding
      resourceRef: thetranslator5
      id: zzz
      envVarPrefix: qiye
```

Check the secret `binding-request-sample` has been injected:

```
kubectl get appconfigs.stable.example.com -o yaml
apiVersion: v1
items:
- apiVersion: stable.example.com/v1
  kind: AppConfig
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"stable.example.com/v1","kind":"AppConfig","metadata":{"annotations":{},"name":"demo-appconfig","namespace":"default"},"spec":{"Command":"some command","image":"my-image","spec":{"containers":[{"image":"yusufkaratoprak/kubernetes-gosample:latest","name":"hello-world","ports":[{"containerPort":8090}]}]},"uri":"some uri"}}
    creationTimestamp: "2020-04-30T01:07:21Z"
    generation: 12
    name: demo-appconfig
    namespace: default
    resourceVersion: "4909355"
    selfLink: /apis/stable.example.com/v1/namespaces/default/appconfigs/demo-appconfig
    uid: e0175a4c-a442-4b7c-a2c7-de132eb4dd7a
  spec:
    Command: some command
    image: my-image
    spec:
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-05-22T08:08:51Z"
        envFrom:
        - secretRef:
            name: binding-request-sample
        image: yusufkaratoprak/kubernetes-gosample:latest
        name: hello-world
        ports:
        - containerPort: 8090
        resources: {}
    uri: some uri
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

Delete the SBR and secret has been removed.

```
spec:
      containers:
      - env:
        - name: ServiceBindingOperatorChangeTriggerEnvVar
          value: "2020-05-22T08:08:51Z"
        image: yusufkaratoprak/kubernetes-gosample:latest
        name: hello-world
        ports:
        - containerPort: 8090
        resources: {}
```

File: pkg/controller/servicebindingrequest/binder_test.go [hunk 8]
```diff
@@ -184,6 +244,17 @@ func TestBinderNew(t *testing.T) {
 		// making sure no volume mounts are present
 		require.Nil(t, c.VolumeMounts)
 	})
+
```
Comment (isutton) [inline pkg/controller/servicebindingrequest/binder_test.go:247]: Is it possible to expand on what is being tested here?

**Comment by sbose78** [comment]
replaced by #513

**Pull request closed without merging**